### ISSUE TITLE
Completion of function calls sometimes fails to include snippet

### ISSUE THREAD

**Issue by HighCommander4** [clangd/clangd#1785]
In the following code:

```c++
void waldo(int arg);

int main()
{
  int val;
  wal^

    // if (true)
  
}
```

Place the cursor at the location indicated by `^`, invoke completion, and select the only result.

Expected result: `waldo(int arg)` is inserted.

Actual result: `waldo` is inserted.

The behaviour has something to do with the commented line.

**Comment by HighCommander4** [comment]
[This branch](https://searchfox.org/llvm/rev/a502dddfd0da8ccefd2cee15599b49f6eaf74efa/clang-tools-extra/clangd/CodeComplete.cpp#597) is being incorrectly taken, because `NextTokenKind` is incorrectly calculated to be `tok::l_paren`.

**Comment by HighCommander4** [comment]
`NextTokenKind` is computed [here](https://searchfox.org/llvm/rev/a502dddfd0da8ccefd2cee15599b49f6eaf74efa/clang-tools-extra/clangd/CodeComplete.cpp#1592-1596).

Going into `Lexer::findNextToken()`, the input `Loc` is the code completion location, which in the source buffer looks like a null character was inserted in between the source character before the cursor and the source character after the cursor when completion was invoked.

Things go downhill on this line, though:

```c++
  Loc = Lexer::getLocForEndOfToken(Loc, 0, SM, LangOpts);
```

After this, `Loc` now points to the space character before the `(` character in the comment, which explains why the next token from that point onwards is identified as an `l_paren`.

**Comment by HighCommander4** [comment]
Having a bit of a "how did this ever work" moment here.

Inside `getLocForEndOfToken()`, we [call](https://searchfox.org/llvm/rev/a502dddfd0da8ccefd2cee15599b49f6eaf74efa/clang/lib/Lex/Lexer.cpp#813) `MeasureTokenLength()`, which lexes the next non-whitespace token _after_ the code completion location, which is the `tok::comment` token `// if (true)`, and returns _its_ length, which is 12...

... but then adds that 12 to the code completion location (which differs from the start of the comment token by the intervening whitespace that was skipped), resulting in a location in the middle of the comment token.

**Comment by HighCommander4** [comment]
I guess the way the current code manages to successfully identify cases where the next token after the completion location really is a `(` (which is what it's trying to do), is that in those cases, `MeasureTokenLength()` measures the length of that `(` token, which is 1, and then adds that 1 to the code completion location (which points to the null character), resulting in the offset where we actually do want to start lexing for the next token.

It basically works by accident, but the mechanism is pretty badly broken.

**Comment by HighCommander4** [comment]
One more thing I noticed while stepping through the lexer code that may be relevant: the lexer's inner loop does have a [check](https://searchfox.org/llvm/rev/a502dddfd0da8ccefd2cee15599b49f6eaf74efa/clang/lib/Lex/Lexer.cpp#3659) for the code completion location, but checking for this requires the lexer to have a `Preprocessor`, and the temporary lexer created inside `Lexer::getRawToken()` (called from `MeasureTokenLength()`) does not have one.

**Comment by HighCommander4** [comment]
My best idea for how to fix this is to have clangd not use `Lexer::findNextToken`, but instead use code similar to the implementation of `findNextToken`, but which works for the completion point's `SourceLocation`.

It would be nicer to get `findNextToken` itself to handle the completion point, but I don't see how it can, given that the completion point is stored on the `Preprocessor`, and `findNextToken`'s arguments do not include anything from which you can recover the `Preprocessor`; we'd have to modify the interfaces of `findNextToken`, `getLocForEndOfToken`, and `MeasureTokenLength` to take a `Preprocessor` as an optional parameter, or something like that.

**Comment by zyn0217** [comment]
Thanks for the fix. Looking around the changed lines, I remembered that once I encountered the case where the "don't add new parens & placeholders if arg list does exist." heuristic doesn't work when completing a setter function call generated by Protobuf. I'd like to have a try on this patch and see if this solves that problem. (I didn't get around to reducing the code snippet since I first ran into it. Looks like it is somehow particular to protobuf because I've never seen this issue for my hand-written functions.)

![](https://github.com/clangd/clangd/assets/117258288/aa707b70-f73f-4bf7-bb0e-d1c4f4cf1544)
